File: android/java/org/chromium/chrome/browser/PlaylistPageHandlerFactory.java
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

package org.chromium.chrome.browser;

import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.crypto_wallet.util.Utils;
import org.chromium.chrome.browser.profiles.Profile;
import org.chromium.mojo.bindings.ConnectionErrorHandler;
import org.chromium.mojo.bindings.Interface;
import org.chromium.mojo.bindings.Interface.Proxy.Handler;
import org.chromium.mojo.system.MessagePipeHandle;
import org.chromium.mojo.system.impl.CoreImpl;
import org.chromium.playlist.mojom.PageHandler;

@JNINamespace("chrome::android")
public class PlaylistPageHandlerFactory {
    private static final Object lock = new Object();
    private static PlaylistPageHandlerFactory instance;

    public static PlaylistPageHandlerFactory getInstance() {
        synchronized (lock) {
            if (instance == null) {
                instance = new PlaylistPageHandlerFactory();
            }
        }
        return instance;
    }

    private PlaylistPageHandlerFactory() {}

    public PageHandler getPlaylistPageHandler(ConnectionErrorHandler connectionErrorHandler) {
        Profile profile = Utils.getProfile(false); // Always use regular profile
        if (profile == null) {
            return null;
        }
        int nativeHandle =
                PlaylistPageHandlerFactoryJni.get().getInterfaceToPlaylistPageHandler(profile);
        if (nativeHandle == -1) {
            return null;
        }
        MessagePipeHandle handle = wrapNativeHandle(nativeHandle);
        PageHandler pageHandler = PageHandler.MANAGER.attachProxy(handle, 0);
        Handler handler = ((Interface.Proxy) pageHandler).getProxyHandler();
        handler.setErrorHandler(connectionErrorHandler);

        return pageHandler;
    }

    private MessagePipeHandle wrapNativeHandle(int nativeHandle) {
        return CoreImpl.getInstance().acquireNativeHandle(nativeHandle).toMessagePipeHandle();
    }

    @NativeMethods
    interface Natives {
        int getInterfaceToPlaylistPageHandler(Profile profile);
    }
}

File: browser/playlist/android/playlist_page_handler_factory.cc
/* Copyright (c) 2022 The Brave Authors. All rights reserved.
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this file,
 * You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "brave/browser/playlist/android/playlist_page_handler_factory.h"

#include <utility>

#include "brave/browser/brave_browser_process.h"
#include "brave/browser/playlist/playlist_service_factory.h"
#include "chrome/browser/browser_process.h"
#include "chrome/browser/profiles/incognito_helpers.h"
#include "chrome/browser/profiles/profile.h"
#include "components/keyed_service/content/browser_context_dependency_manager.h"

namespace playlist {

// static
PlaylistPageHandlerFactory* PlaylistPageHandlerFactory::GetInstance() {
  return base::Singleton<PlaylistPageHandlerFactory>::get();
}

// static
mojo::PendingRemote<mojom::PageHandler>
PlaylistPageHandlerFactory::GetForContext(content::BrowserContext* context) {
  return static_cast<PlaylistAndroidPageHandler*>(
             GetInstance()->GetServiceForBrowserContext(context, true))
      ->MakeRemote();
}

PlaylistPageHandlerFactory::PlaylistPageHandlerFactory()
    : BrowserContextKeyedServiceFactory(
          "PlaylistAndroidPageHandler",
          BrowserContextDependencyManager::GetInstance()) {
  DependsOn(playlist::PlaylistServiceFactory::GetInstance());
}

PlaylistPageHandlerFactory::~PlaylistPageHandlerFactory() = default;

KeyedService* PlaylistPageHandlerFactory::BuildServiceInstanceFor(
    content::BrowserContext* context) const {
  auto* profile = Profile::FromBrowserContext(context);
  playlist::PlaylistServiceFactory::GetInstance()->GetForBrowserContext(
      context);
  return new PlaylistAndroidPageHandler(profile);
}

content::BrowserContext* PlaylistPageHandlerFactory::GetBrowserContextToUse(
    content::BrowserContext* context) const {
  return chrome::GetBrowserContextRedirectedInIncognito(context);
}

}  // namespace playlist